rust/tool/src/esp.rs

```diff
@@ -1,20 +1,62 @@
 use std::array::IntoIter;
 use std::path::{Path, PathBuf};
+use std::str::FromStr;
 
-use anyhow::{Context, Result};
+use anyhow::{Context, Result, Error, bail};
 
 use crate::generation::Generation;
 
+/// Supported system
+#[allow(dead_code)]
+#[non_exhaustive]
+pub enum System {
+    X86,
+    AArch64,
+}
+
+impl System {
+    pub fn systemd_stub_filename(&self) -> &Path {
+        Path::new(match self {
+            Self::X86 => "linuxx64.efi.stub",
+            Self::AArch64 => "linuxaa64.efi.stub"
+        })
+    }
+
+    pub fn systemd_filename(&self) -> &Path {
+        Path::new(match self {
+            Self::X86 => "systemd-bootx64.efi",
+            Self::AArch64 => "systemd-bootaa64.efi"
+        })
+    }
+
+    pub fn efi_fallback_filename(&self) -> &Path {
+        Path::new(match self {
+            Self::X86 => "BOOTX64.EFI",
+            Self::AArch64 => "BOOTAA64.EFI",
+        })
+    }
+}
+
+impl FromStr for System {
```

Comment:
Instead of `FromStr` we could just implement a `from_nixos_system` function. This way it is clear that we are constructing an architecture single from the nixos system double.